README.md
# Arras Template, Made by 
This is a heavily optimized and improved upon version of https://github.com/nepphhh/arrasio.<br>
As such, it's very different from the source code.<br>
The client and server are all split up into many different files.
Each file's name should give you a description of what it does.
## Setting up a static client + server on one project
1) In the server's `config.json`, find the `host` property and set it to the proper host (eg. `localhost`, `arras.io`, `your-project.herokuapp.com`, `your-thing.glitch.me`).
2) Make sure the `port` property is correct in the `config.json` file.<br> Please note that if you're using a hosting solution that sets the port for you, it will be the value of `process.env.PORT`. In `global.js`, this is taken care of for you.
3) `npm start` or `node server/index.js`
## Setting up a standalone client somewhere else
1) Make sure you have the `standaloneClient` folder.
2) In the `index.js` file of that folder, make sure the `ip` property on line `9` is set to the complete host and port of the server you want the client to connect to. (eg `127.0.0.1:3000`, `234.23.123.123:3001`, `your-project.herokapp.com`)
3) `npm run host`
## Setting up a standalone server with a standalone client
1) Make sure the steps above are complete
2) In the server's `config.json`, find the `host` property and set it to the proper host (eg. `localhost`, `arras.io`, `your-project.herokuapp.com`, `your-thing.glitch.me`).
3) Make sure the `port` property is correct in the `config.json` file.<br> Please note that if you're using a hosting solution that sets the port for you, it will be the value of `process.env.PORT`. In `global.js`, this is taken care of for you.
4) Make sure the `servesStatic` property is set to `false` in the `config.json` file.
5) `npm start` or `node server/index.js`

public/changelog.md
# Update (2/17/2023)
- Buffed pounder,destroyer,annihilator bullet health,pen,damage
- Fixed the discord invite
- Made the gamespeed a bit faster

# Update (2/16/2023)
- Upgrade Basic Bug Fixed
- Food Rarety Per Tier

# Update (2/15/2023)
- Better Testbed Setup
- Cursed Tank
- Portal

# Plans (2/15/2023)
- Map size of 4000x16000
- Rooms for each 4000x4000
- Portal (Added)
- Food to spawn in Respective rooms

# Update (2/15/2023)
- Reworked Food Score now it quadruples per level,
- Square XP: 107
- Triangle XP: 427
- Pentagon XP: 1717

# Update (2/11/2023)
- Added Egger

# Update (2/2/2023)
- Removed skim turret from Glitchton minions.

# Update (31/1/2023)
- Added New Logo (pecoka)

# Update(28/1/2023)
- Added Elite Spinner from arras.io

# Update (24/1/2023)
- Added Theia
- Added Nyx
- Now Beta Testers have access to Semi-Celestials

# Update (24/1/2023)
- Added Square Cluster,
- Food can colide with themselves

# Update (23/1/2023)
- Remade arras.io Big Cheese tank

# Update (21/1/2023)
- Nerfed Ganymede 2nd layer reload speed

# Update (21/1/2023)
- Added Ridan (Creidits to !Attakante#8485)
- Added Ridan's ling celestial (Ganymede)

# Update (16/1/2023)
- Added Trimurti

# Update (1/2/2023)
- Siege Bots count is now 22
- Bosses will loose health in boss room insted of regening it on damage
- Added Glitchton (Credits to Lavamani#6876)

# Update (12/30/2022)
- Added Pentamancer
- Added some helpers to deal with bosses (big smashers)
- Added Elite Rifle

# Update (12/25/2022)
- Christmas update!
- Added spiky (Spawns as obstacles)
- For who that thinks spiky will make siege harder
- Added Cutter Spawner (This one spawns a giant cutter with lot of health)

# Update (12/18/2022)
- Changed sanctuary design with a flank turret for more defense

# Update (12/18/2022)
- Nerfed The Disease minions.
- Added Magmaporous
- Added Legionary Crasher
- Added Zamolxis (A Semi Celestial)

# Update(12/17/2022)
- Added The Disease (Credits to Dwajl)
- A Boss with celestial health and body damage but all its layers has 8 sides (insted of themain body which has 12 sides)
- first layer has 4 sunchip spawners that deal 2x damage
- 2nd has 4 twin turrets and 4 ranger turrets
- and final has 8 factory guns
- if there are equal or over 7 
  it will be class 5 difficulty (hard) 
  if over 5
  or equal
  it will be class 4 difficulty (Intermediete)
  if over or equal to 3
  it will be class 2 difficulty (Novice)
  if 1 or 2
  it will be class 1 difficulty (Easy)

# Update (12/17/2022)
- Bosses now orbit around sanctuaries at further distance to let Eternals not ram the sanctuaries.
- Added Xumes, a new Celestial.

# Update (12/16/2022)
- Bots name are now Wanderer (Number between 1 and 54533)

# Update (12/16/2022)
- Bot movement changed.
- Added Eternal waves from 14 to 25

# Update (12/11/2022)
- Maze 4TDM Bots increased from 24 to 48
- Added a new celestial Enyo

# Update (12/10/2022)
- Added Kronos
- Added Zaphkiel. Paladin, Freyja

# Startup (12/10/2022)
- Server Startup.

public/lib/gameDraw.js
import { util } from "./lib/util.js";
import { global } from "./lib/global.js";
import { config } from "./lib/config.js";
import { Canvas } from "./lib/canvas.js";
import { color } from "./lib/color.js";
const drawingFunctions = (function (util, global, config, Canvas, color) {
  let mixColors = (() => {
    /** https://gist.github.com/jedfoster/7939513 **/
    function d2h(d) {
      return d.toString(16);
    } // convert a decimal value to hex
    function h2d(h) {
      return parseInt(h, 16);
    } // convert a hex value to decimal
    return (color_2, color_1, weight = 0.5) => {
      if (weight === 1) return color_1;
      if (weight === 0) return color_2;
      var col = "#";
      for (var i = 1; i <= 6; i += 2) {
        // loop through each of the 3 hex pairs—red, green, and blue, skip the '#'
        var v1 = h2d(color_1.substr(i, 2)), // extract the current pairs
          v2 = h2d(color_2.substr(i, 2)),
          // combine the current pairs from each source color, according to the specified weight
          val = d2h(Math.floor(v2 + (v1 - v2) * weight));
        while (val.length < 2) {
          val = "0" + val;
        } // prepend a '0' if val results in a single digit
        col += val; // concatenate val to our new color string
      }
      return col; // PROFIT!
    };
  })();
  function getRainbow(a, b, c = 0.5) {
    if (0 >= c) return a;
    if (1 <= c) return b;
    let f = 1 - c;
    a = parseInt(a.slice(1, 7), 16);
    b = parseInt(b.slice(1, 7), 16);
    return (
      "#" +
      (
        (((a & 16711680) * f + (b & 16711680) * c) & 16711680) |
        (((a & 65280) * f + (b & 65280) * c) & 65280) |
        (((a & 255) * f + (b & 255) * c) & 255)
      )
        .toString(16)
        .padStart(6, "0")
    );
  }
  function getColor(colorNumber) {
    // 0-24, 30-37 and 39-41 are kept the same as standard Arras for consistency.
    // For Woomy-Arras color IDs between 20 and 44, increment the number by 30 so they'll work as intended.

    // The following Woomy-Arras color IDs are not included as they are duplicates of another color.
    /*
    Original Color ID -> Replacement Color ID
    68 -> 108
    226 -> 13
    */
    // Other Woomy-Arras color IDs will work fine with no changes.

    // The following Raxor.io color IDs are not included as they are duplicates of another color.
    /*
    Original Color ID -> Replacement Color ID
    43 -> 190
    */
    switch (colorNumber) {
      case 0:
        return color.teal;
      case 1:
        return color.lgreen;
      case 2:
        return color.orange;
      case 3:
        return color.yellow;
      case 4:
        return color.lavender;
      case 5:
        return color.pink;
      case 6:
        return color.vlgrey;
      case 7:
        return color.lgrey;
      case 8:
        return color.guiwhite;
      case 9:
        return color.black;
      case 10:
        return color.blue;
      case 11:
        return color.green;
      case 12:
      case "FFA_RED":
        return color.red;
      case 13:
        return color.gold;
      case 14:
        return color.purple;
      case 15:
        return color.magenta;
      case 16:
        return color.grey;
      case 17:
        return color.dgrey;
      case 18:
        return color.white;
      case 19:
        return color.guiblack;
      case 20:
        return 150 > Date.now() % 300 ? color.blue : color.red;
      case 21:
        return 150 > Date.now() % 300 ? color.blue : color.grey;
      case 22:
        return 150 > Date.now() % 300 ? color.grey : color.blue;
      case 23:
        return 150 > Date.now() % 300 ? color.red : color.grey;
      case 24:
        return 150 > Date.now() % 300 ? color.grey : color.red;
      case 30:
        return "#a913cf";
      case 31:
        return "#226ef6";
      case 32:
        return "#ff1000";
      case 33:
        return "#ff9000";
      case 34:
        return "#00e00b";
      case 35:
        return "#ffd300";
      case 36:
        return getRainbow(
          "#ff1000 #ff9000 #ffd300 #00e00b #226ef6 #a913cf".split(" ")[
            Math.floor((Date.now() / 200) % 6)
          ],
          "#ff9000 #ffd300 #00e00b #226ef6 #a913cf #ff1000".split(" ")[
            Math.floor((Date.now() / 200) % 6)
          ],
          (Date.now() / 200) % 1
        );
      case 37:
        return getRainbow(
          "#ffffff",
          2e3 > Date.now() % 4e3 ? "#55cdfc" : "#f7a8b8",
          5 * Math.sin(((Date.now() % 2e3) / 2e3) * Math.PI) - 2
        );
      case 39:
        return "#654321";
      case 40:
        return "#e58100";
      case 41:
        return "#267524";
      default:
        return "#00000000";
    }
  }
  function getColorDark(givenColor) {
    let dark = config.graphical.neon ? color.white : color.black;
    if (config.graphical.darkBorders) return dark;
    return mixColors(givenColor, dark, color.border);
  }
  function getZoneColor(cell, real) {
        switch (cell) {
            case 'bas1':
            case 'bap1':
            case 'dom1':
            case 'hyou':
                return color.blue;
            case 'bas2':
            case 'bap2':
            case 'dom2':
                return color.green;
            case 'bas3':
            case 'bap3':
            case 'dom3':
            case 'boss':
                return color.red;
            case 'bas4':
            case 'bap4':
            case 'dom4':
                return color.pink;
            case 'nest':
                return (real) ? color.purple : color.lavender;
            case 'cent':
                return "#ff0000"
            case 'dom0':
                return color.gold;
            default:
                return (real) ? color.white : color.lgrey;
        }
    }

  function setColor(context, givenColor) {
    if (config.graphical.neon) {
      context.fillStyle = getColorDark(givenColor);
      context.strokeStyle = givenColor;
    } else {
      context.fillStyle = givenColor;
      context.strokeStyle = getColorDark(givenColor);
    }
  }
  return {
    mixColors,
    getColor,
    getColorDark,
    getZoneColor,
    setColor,
  };
})(util, global, config, Canvas, color);
